dbms/src/Operators/ConcatSourceOp.h

```diff
@@ -0,0 +1,229 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Common/Exception.h>
+#include <Flash/Pipeline/Exec/PipelineExec.h>
+#include <Flash/Pipeline/Exec/PipelineExecBuilder.h>
+#include <Operators/Operator.h>
+
+#include <memory>
+
+namespace DB
+{
+class SetBlockSinkOp : public SinkOp
+{
+public:
+    SetBlockSinkOp(
+        PipelineExecutorStatus & exec_status_,
+        const String & req_id,
+        Block & res_)
+        : SinkOp(exec_status_, req_id)
+        , res(res_)
+    {
+    }
+
+    String getName() const override
+    {
+        return "SetBlockSinkOp";
+    }
+
+protected:
+    OperatorStatus writeImpl(Block && block) override
+    {
+        if unlikely (!block)
+            return OperatorStatus::FINISHED;
+
+        assert(!res);
+        res = std::move(block);
+        return OperatorStatus::NEED_INPUT;
+    }
+
+private:
+    Block & res;
+};
+
+/// Used to merge multiple partitioned tables of storage layer.
+class ConcatSourceOp : public SourceOp
+{
+public:
+    ConcatSourceOp(
+        PipelineExecutorStatus & exec_status_,
+        const String & req_id,
+        std::vector<PipelineExecBuilder> & exec_builder_pool)
+        : SourceOp(exec_status_, req_id)
+    {
+        RUNTIME_CHECK(!exec_builder_pool.empty());
+        setHeader(exec_builder_pool.back().getCurrentHeader());
+        for (auto & exec_builder : exec_builder_pool)
+        {
+            exec_builder.setSinkOp(std::make_unique<SetBlockSinkOp>(exec_status_, req_id, res));
+            exec_pool.push_back(exec_builder.build());
+        }
+    }
+
+    String getName() const override
+    {
+        return "ConcatSourceOp";
+    }
+
+    // ConcatSourceOp is used to merge multiple partitioned tables of storage layer, so override `getIOProfileInfo` is needed here.
+    IOProfileInfoPtr getIOProfileInfo() const override { return IOProfileInfo::createForLocal(profile_info_ptr); }
+
+protected:
+    void operatePrefixImpl() override
+    {
+        if (!popExec())
+            done = true;
+    }
+
+    void operateSuffixImpl() override
+    {
+        if (cur_exec)
+        {
+            cur_exec->executeSuffix();
+            cur_exec.reset();
+        }
+        exec_pool.clear();
+    }
+
+    OperatorStatus readImpl(Block & block) override
+    {
+        if unlikely (done)
+            return OperatorStatus::HAS_OUTPUT;
```

Comment:
Why not return `OperatorStatus::FINISHED`

Reply:
Because of  https://github.com/pingcap/tiflash/blob/f8943e7f3d460feefe98d00b14182b5186d9372b/dbms/src/Operators/Operator.h#L117-L120